CIP-????/README.md

```diff
@@ -0,0 +1,113 @@
+An object can be one of a set of predefined set of entities, distinguished by constructor ids. I.e. the optional Redeemer type is either `Some Redeemer` or `None`.
+Because we know that anything of optional integer type can be either two of these, only two numbers (0/1) are required to distinguish them.
+If we introduce a third constructor (i.e. `Some Datum`), potentially all other constructors change and the two implementations are not compatible anymore.
+
+Moreover there are other Plutus language frontends that allow freely declaring objects and mixing them into Union types (such as OpShin), which is akin to the imperative style of declaring classes.
+This allows for example to declare a universally accepted type `Nothing` that can be freely mixed with `Redeemer` and `Datum` into `Union[Nothing, Redeemer, Datum]`.
+The only requirement to ensure that this works properly is that all records that are mixed into the Union have distinct contstructor ids.
+This is currently implemented manually, which is tedious and a potential source of errors.
+
+## Specification
+<!-- The technical specification should describe the proposed improvement in sufficient technical detail. In particular, it should provide enough information that an implementation can be performed solely on the basis of the design in the CIP. This is necessary to facilitate multiple, interoperable implementations. This must include how the CIP should be versioned. If a proposal defines structure of on-chain data it must include a CDDL schema in it's specification.-->
+The deterministic, universal and almost-unique Plutus constructors are computed recursively based on the type definition of a record.
+We first compute a string `ustr(X)` based on the type definition of X. Then we perform a sha256 hash on the UTF8 encoding of this string and interpret the resulting hex digest as a big endian encoded integer.
+The integer is taken modulo 2^32. The resulting integer is the almost-unique, universal, deterministic constructor id of the plutus datum.
+
+The following function describes how to compute `ustr(X)` for a type recursively.
+
+```
+ustr(bytes) := "bytes"
+ustr(integer) := "int"
+// This covers the case where the structure of the object is now known from the perspective of the class, i.e. when any BuiltinData is allowed
+ustr(PlutusData) := "any"
+  // This covers the case where the type of the elements in the list are not known in advance
+ustr(list) := "list"
+
+ustr(list<X>) := "list<" + ustr(X) + ">
+
+ustr(map<X,Y>) := "map<" + ustr(X) + "," + ustr(Y) + ">"
+
+ustr(union<X,Y,...,Z>) := "union<" + ustr(X) + "," + ustr(Y) + "," + ... + "," + ustr(Y) + ">"
+
+ustr(constr(name)<id, fields[f1:X,f2:Y,...,fn:Z]>) :=
+    "cons[" + name + "](" + str(id) + ";"
+    + f1 + ":" + ustr(X) + "," + f2 + ":" + ustr(Y) + "," + ... + "," + fn + ":" + ustr(Z) +
+    ")"
+```
+
+Where `name` and `f1` to `fn` refer to the name of the record and the names of its fields respectively.
```

Comment:
> Commented later, but: you have to have either the name or the id of the constructor, otherwise you can't distinguish two constructors with the same fields. And the whole point of this proposal is to not set the ids, so it has to be names.

I mean, you can get the hash of a constructor from the structure of the data type (`\a -> Either a ()` for both the example types) and the id of the constructor.  You don't need to globally specify the id this way, just for the purpose of computing a hash (and the way hashes are computed can be arbitrary as long as they're almost unique). Or am I misunderstanding it?